`Development`: Extend cypress e2e documentation (#6996)
ls1intum/Artemis: docs/dev/cypress.rst

    export const USER_ID = {
      studentOne: 1,
      studentTwo: 2,
      studentThree: 3,
      instructor: 16,
      tutor: 6,
    };

4. Open Cypress browser

  If you want to use a different browser than chrome, you can set this within the ``package.json`` file
  within the cypress subfolder like this ``"cypress:open": "cypress open --browser=edge",``.
  To now run the test suites selectively instead of in full, we need to open the cypress
  browser, which is by default chrome by running the following command ``npm run cypress:open``.
  Now select ``E2E Testing``, followed by ``Start E2E testing in ...``. A new browser window
  should open, which should look like this:

  .. figure:: cypress/cypress-open-screenshot.png
    :align: center
    :alt: Cypress cypress-open-screenshot

  You can now click on any test suite and it should run. 
  **IMPORTANT**: If you run the E2E tests for the first time, always run the ``ImportUsers.ts`` tests first,
  since it will create the necessary users.

Debug using Sorry Cypress
-------------------------
In order to debug failing E2E Cypress tests with a dashboard (called Sorry Cypress), please follow the instructions

> on confluence: https://confluence.ase.in.tum.de/display/ArTEMiS/Sorry+Cypress+Dashboard

Reviewer: Shouldn't we put this into the documentation directly (except for the credentials of course)?

Author: Good idea, done!

> Best Practice when writing new E2E tests

Reviewer: ```suggestion
Best practice when writing new E2E tests
```

consistency

Therefore, the current setup only dynamically deploys the Artemis server and configures it to connect to
the prelive system, which is already properly setup in the university data center.

Local Cypress Setup
-------------------

> Sometimes developers need to setup Cypress locally, in order to debug failing E2E tests or write new tests.
> In order to setup cypress locally, we need to follow these steps:

Reviewer 2: https://grammarist.com/spelling/set-up-vs-setup/
Also you used "set up" twice:
```suggestion
Sometimes developers need to set up Cypress locally, in order to debug failing E2E tests or write new tests.
Follow these steps to create your local cypress instance:
```

1. Install dependencies

  First head into the cypress folder by using ``cd /src/test/cypress``. Now run ``npm install``.

2. Customize Cypress settings

  To connect cypress to our local Artemis instance, we need to adjust some configurations. 
  First we need to set the URL or IP of the Artemis instance in the ``cypress.config.ts`` file.
  Adjust the ``baseUrl`` setting to fit your setup (e.g. ``baseUrl: 'http://localhost:9000',``)

3. Adjust user settings

  We also need to adjust the user setting, which will determine the usernames and passwords, that cypress
  will use. These settings are located within the ``cypress.env.json`` file. If you use the Atlassian setup,
  the file should typically look like this:
   
  .. code-block:: json

    {
      "username": "artemis_test_user_USERID",
      "password": "artemis_test_user_USERID",
      "adminUsername": "artemis_admin",
      "adminPassword": "artemis_admin",
      "allowGroupCustomization": true,
      "studentGroupName": "students",
      "tutorGroupName": "tutors",
      "editorGroupName": "editors",
      "instructorGroupName": "instructors"
    }

>   The ``USERID`` part will be replaced by different user ids. These are set within the ``support/users.ts`` file. 

Reviewer 2: To strengthen that cypress will do this and you don't need to change that manually in the file:
```suggestion
  The ``USERID`` part will be automatically replaced by different user ids. These are set within the ``support/users.ts`` file. 
```

  Now select ``E2E Testing``, followed by ``Start E2E testing in ...``. A new browser window
  should open, which should look like this:

  .. figure:: cypress/cypress-open-screenshot.png
    :align: center
    :alt: Cypress cypress-open-screenshot

  You can now click on any test suite and it should run. 

.. warning::
  **IMPORTANT**: If you run the E2E tests for the first time, always run the ``ImportUsers.ts`` tests first,
  since it will create the necessary users.

Debug using Sorry Cypress
-------------------------

Since the E2E tests are sometimes hard to debug, we provide a dashboard, that allows to inspect the 
CI run and even watch a video of the UI interaction with Artemis in that run. 

It's based on Sorry Cypress a open source and selfhostable alternative to the paid cypress cloud.

The dashboard itself can be access here: https://sorry-cypress.ase.cit.tum.de/

To access it, you need these basic auth credentials (sorry cypress itself does not provide an auth 
system, so we are forced to use nginx basic auth here). You can find these credentials on our confluence page:
https://confluence.ase.in.tum.de/display/ArTEMiS/Sorry+Cypress+Dashboard

> After that you will see the initial dashboard without any useful information. 

Reviewer 2: ```suggestion
After that you will see the initial dashboard. 
```

You first have to select a project in the left sidebar (mysql or postgresql):

  .. figure:: cypress/sorry-cypress-dashboard.png
    :align: center
    :alt: Sorry Cypress dashboard

Now you get a list of the last runs. In the top right you can enter your branch name to filter the runs.

  .. figure:: cypress/sorry-cypress-runs.png
    :align: center
    :alt: Sorry Cypress last runs

> The name of the run consists of the branch name followed by the run # . The last part is MySQL or 

Reviewer 2: ```suggestion
The name of the run consists of the branch name followed by the run number. The last part is MySQL or 
```

> PostgreSQL depending on the run environment. If you are in the MySQL project, you will of course only see the MySQL  runs. 

Reviewer 2: ```suggestion
PostgreSQL depending on the run environment. If you are in the MySQL project, you will of course only see the MySQL runs. 
```
